Control nesting level for shards skipping and disallow non-deterministic functions

azat opened the pull request:
Yet another small improvements around distributed querying.

Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
- Add `optimize_skip_unused_shards_nesting` (allows control nesting level for shards skipping optimization)
- Add `force_skip_optimize_shards_nesting` (allows control nesting level for checking was shards skipped or not)
- Deprecate `force_optimize_skip_unused_shards_no_nested` (`force_skip_optimize_shards_nesting` should be used instead)
- Disable `optimize_skip_unused_shards` if sharding_key has non-deterministic func (i.e. `rand()`, note that this does not changes anything for INSERT side)

<details>

HEAD:
- 080e309050ee844db8cf70951825da9e41f7e987
- 0e218b0f15a502ed5c95499e13fa89f8d52fa006

</details>

azat committed:
Disable optimize_skip_unused_shards if sharding_key has non-deterministic func

Example of such functions is rand()

And this patch disables only optimize_skip_unused_shards, i.e. INSERT
code path does not changed, so it will work as before.

azat committed:
optimize_skip_unused_shards=2 will disable it for nested distributed queries

P.S. Looks like settings can be converted between SettingUInt64 and
SettingBool without breaking binary protocol.

FWIW maybe it is a good idea to change the semantics of the settings as
follow (but I guess that changing semantic is not a good idea, better to
add new settings and deprecate old ones):
- optimize_skip_unused_shards -- accept nesting level on which the
  optimization will work
- force_skip_optimize_shards_nesting -- accept nesting level on which
  the optimization will work

azat committed:
Add missing DROP TABLE in 01319_mv_constants_bug

azat committed:
Add logging of adjusting conditional settings for distributed queries

azat committed:
Improve 01319_optimize_skip_unused_shards_no_nested

Before there is no check that optimize_skip_unused_shards was working
for the first level, use cluster with unavalable shard to guarantee
this.

alexey-milovidov commented:
> P.S. Looks like settings can be converted between SettingUInt64 and
SettingBool without breaking binary protocol?

Yes.

alexey-milovidov commented:
> 
    optimize_skip_unused_shards_nesting -- accept nesting level on which the
    optimization will work
    force_skip_optimize_shards_nesting -- accept nesting level on which
    the optimization will work

Much better than `optimize_skip_unused_shards=2`

alexey-milovidov reviewed (approved):
The code LGTM,
let's apply changes to make settings more convenient.

alexey-milovidov commented:
BTW, what is the setup when nested Distributed tables are required?
(the question from Yandex Meetup)

azat committed:
Add settings to control nesting level for shards skipping optimization

- optimize_skip_unused_shards_nesting (allows control nesting level for
  shards skipping optimization)
- force_skip_optimize_shards_nesting (allows control nesting level for
  checking was shards skipped or not)
- deprecates force_skip_optimize_shards_no_nested

azat commented:
>BTW, what is the setup when nested Distributed tables are required?

I was going to add some block about use case into the documentation (since there can be some interesting use cases and also caveats), but did not manage to find time for this, egh.

Anyway let me try to explain it briefly here:

Suppose you have lots of nodes (say 1000) in the cluster.
So what can you do in this case is divide your cluster of 1000 nodes, into smaller, for simplicity let's use 100 nodes per cluster, so 10 smaller cluster of 100 nodes. And the distribute data using some greatly-distributed-column (let's say user_id) between smaller clusters, and for this you will have two distributed tables, one that will distribute between smaller clusters (with sharding_key = consistent hash(user_id)), and distribute table in each cluster (let's use random distribution there for simplicity).
Then when you need to query your data you can use `optimize_skip_unused_shards` to avoid sending queries to all 1000 nodes, only to 100.

And one more advantage rise up when you need to expand the cluster (basically all above can be solved without nesting, but will require some trickery), since in this case you can just add new nodes into smaller cluster without any data re-sharding.

P.S. note that this a brief descriptions, that does not accounts some possible issues/aspects.

alexey-milovidov commented:
> Yandex synchronization check (only for Yandex employees) 

This is Ok.

alexey-milovidov commented:
```
/build/build_docker/../src/Interpreters/ClusterProxy/executeQuery.cpp:55:56: error: converting integer literal to bool, use bool literal instead [modernize-use-bool-literals,-warnings-as-errors]
            new_settings.optimize_skip_unused_shards = 0;
                                                       ^
                                                       false
```

alexey-milovidov committed:
Update executeQuery.cpp

azat committed:
Fix documentation

azat commented on src/Storages/StorageDistributed.cpp:
```diff
@@ -700,6 +714,8 @@ ClusterPtr StorageDistributed::getOptimizedCluster(const Context & context, cons
         std::stringstream exception_message;
         if (!has_sharding_key)
             exception_message << "No sharding key";
+        else if (sharding_key_is_deterministic)
```
Sorry, there is a typo, here `!sharding_key_is_deterministic` should be.
@alexey-milovidov can you commit this simple fix?

azat replied:
@alexey-milovidov friendly ping
